extendr-api/src/robj/try_from_robj.rs

```diff
@@ -2,54 +2,95 @@
+                    _ => return Err(Error::ExpectedScalar(robj)),
+                };
+
+                // Check if the value is not a missing value
+                if robj.is_na() {
+                    return Err(Error::MustNotBeNA(robj));
+                }
+
+                // If the conversion is int-to-int, check the limits. This
+                // needs to be down by `TryFrom` because the conversion by `as`
+                // is problematic on unsigned integer types.
+                if let Some(v) = robj.as_integer() {
+                    if let Ok(v) = Self::try_from(v) {
+                        return Ok(v);
+                    } else {
+                        return Err(Error::OutOfLimits(robj));
                     }
-                } else if let Some(v) = robj.as_real_slice() {
-                    match v.len() {
-                        0 => Err(Error::ExpectedNonZeroLength(robj)),
-                        1 => {
-                            if !v[0].is_na() {
-                                Ok(v[0] as Self)
-                            } else {
-                                Err(Error::MustNotBeNA(robj))
-                            }
-                        }
-                        _ => Err(Error::ExpectedScalar(robj)),
+                }
+
+                // If the conversion is float-to-int, check if the value is
+                // integer-ish. This needs to be down with `as`, as no `TryFrom`
+                // is implemented for float types. `FloatToInt` trait might
+                // eventually become available in future, though.
+                if let Some(v) = robj.as_real() {
+                    let result = v as Self;
+                    if ((result as f64 - v).abs() < f64::EPSILON) {
+                        return Ok(result);
+                    } else {
+                        return Err(Error::ExpectedIntegerish(robj));
                     }
-                } else {
-                    Err(Error::ExpectedNumeric(robj))
                 }
+
+                Err(Error::ExpectedNumeric(robj))
+            }
+        }
+    };
+}
+
+macro_rules! impl_try_from_scalar_real {
+    ($t:ty) => {
+        impl TryFrom<Robj> for $t {
+            type Error = Error;
+
+            fn try_from(robj: Robj) -> Result<Self> {
+                // Check if the value is a scalar
+                match robj.len() {
+                    0 => return Err(Error::ExpectedNonZeroLength(robj)),
+                    1 => {}
+                    _ => return Err(Error::ExpectedScalar(robj)),
+                };
+
+                // Check if the value is not a missing value
+                if robj.is_na() {
+                    return Err(Error::MustNotBeNA(robj));
+                }
+
+                if let Some(v) = robj.as_integer() {
+                    return Ok(v as Self);
+                }
+                if let Some(v) = robj.as_real() {
+                    return Ok(v as Self);
+                }
```

Comment:
Not sure I understand the logic here (first trying to convert to integer, then to real). Could you add a few comments explaining what is happening?

Reply:
`as_xxx()` can convert only when the underlying SEXP is the same type, so we cannot use `as_real()` directly on INTSXP. Does this answer your question? I might not get your point, sorry.

Reply:
Yes. If you could put something like that into a comment above this section of the code that would be great.

Reply:
I see. I wondered if we should implement such a method with coercion (`Rf_coerceVector`) like `to_real()` on `Robj`'s side, but let's document the current situation for now.